compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/RegExpLexer.java

```diff
@@ -164,62 +125,100 @@ private STToken readTokenInReTerm() {
-            // Handle "[" ["^"] [ReCharSet] "]".
+                return processReEscape();
+            // Start parsing character class [[^] [ReCharSet]].
             case LexerTerminals.OPEN_BRACKET:
                 this.reader.advance();
-                if (peek() == LexerTerminals.BITWISE_XOR) {
-                    startMode(ParserMode.RE_NEGATED_CHAR_CLASS_START);
-                } else {
-                    startMode(ParserMode.RE_CHAR_SET);
-                }
+                startMode(ParserMode.RE_CHAR_SET);
                 return getRegExpSyntaxToken(SyntaxKind.OPEN_BRACKET_TOKEN);
-            // Handle "(" ["?" ReFlagsOnOff ":"] ReDisjunction ")".
+            // Start parsing capturing group ([? ReFlagsOnOff :] ReDisjunction).
             case LexerTerminals.OPEN_PARANTHESIS:
-                if (this.reader.peek(1) == LexerTerminals.QUESTION_MARK) {
-                    startMode(ParserMode.RE_FLAGS_START);
+                this.reader.advance();
+                if (peek() == LexerTerminals.QUESTION_MARK) {
+                    startMode(ParserMode.RE_FLAG_EXPRESSION);
                 } else {
-                    startMode(ParserMode.RE_CAPTURING_GROUP_RE_DISJUNCTION);
+                    startMode(ParserMode.RE_DISJUNCTION);
                 }
-                this.reader.advance();
                 return getRegExpSyntaxToken(SyntaxKind.OPEN_PAREN_TOKEN);
             case LexerTerminals.CLOSE_PARANTHESIS:
                 this.reader.advance();
-                if (this.mode == ParserMode.RE_CAPTURING_GROUP_RE_DISJUNCTION) {
-                    switchMode(ParserMode.RE_QUANTIFIER);
+                // modeStack.size() is > 1 if the close parenthesis is found in a capturing group.
+                // If not it's an invalid token.
+                if (this.modeStack.size() > 1) {
+                    endMode();
+                    startMode(ParserMode.RE_QUANTIFIER);
                     return getRegExpSyntaxToken(SyntaxKind.CLOSE_PAREN_TOKEN);
                 }
-                reportLexerError(DiagnosticErrorCode.ERROR_INVALID_TOKEN_IN_REG_EXP);
-                return getRegExpText(SyntaxKind.RE_CHAR);
+                break;
+            case LexerTerminals.QUESTION_MARK:
+                this.reader.advance();
+                return getRegExpText(SyntaxKind.QUESTION_MARK_TOKEN);
             default:
-                // Handle ReLiteralChar.
                 this.reader.advance();
                 if (!isReSyntaxChar(nextChar)) {
                     startMode(ParserMode.RE_QUANTIFIER);
-                } else {
-                    reportLexerError(DiagnosticErrorCode.ERROR_INVALID_TOKEN_IN_REG_EXP);
+                    return getRegExpText(SyntaxKind.RE_LITERAL_CHAR);
                 }
-                return getRegExpText(SyntaxKind.RE_CHAR);
         }
+
+        reportLexerError(DiagnosticErrorCode.ERROR_INVALID_TOKEN_IN_REG_EXP);
```

Comment:
This error should be logged from the parser side. I think we can just do `return getRegExpText(SyntaxKind.RE_LITERAL_CHAR);` in the default case after adding cases for every syntax char